gh-118633: Add warning regarding the unsafe usage of eval and exec (#118437)
python/cpython: Doc/library/functions.rst

   :returns: The result of the evaluated expression.
   :raises: Syntax errors are reported as exceptions.

   .. warning::

>       This function can execute arbitrary code. Calling it with

Reviewer: Can -> is designed to?

Author: Good point. I will adjust that later today.

Reviewer 2: Or just:
```suggestion
      This function executes arbitrary code. Calling it with
```

Reviewer: The thing in my mind is that this is very different from ast.literal_eval, which has a warning because it's surprising for a function documented to "not be eval therefore doesn't execute arbitrary code" to have a security issue (which is unrelated to executing arbitrary code).

It's very unsurprising that an eval function executes arbitrary code, so the security warning is not something I'd personally think to add... I assume the reason is to remind people why the fact that it executes arbitrary code matters.

I think it's important to avoid accidentally giving the impression that "execute arbitrary code" is a surprising hidden behavior of eval(). It's the kind of thing that would make me afraid upon seeing such a warning that I don't understand understand the function and must surely be missing some reason why it was necessary to add that warning.

Author: I'Ve adapted the text like it was suggested by @JelleZijlstra.

Reviewer: I still believe that this PR is currently incorrect and misleading, and it should say "is designed to" as I suggested above. :)

>       user-supplied input may lead to security vulnerabilities.
> 

Reviewer 3: Maybe directly refer to `ast.literal_eval()` here? You need to remove the sentence further down though.

```suggestion
      user-supplied input may lead to security vulnerabilities.
      See :func:`ast.literal_eval` for a function that can safely evaluate strings
      with expressions containing only literals.

```

Reviewer 4: Let's do this in a future PR if desired.

>       Because this function can be used for executing arbitrary code, it is not recommended
>       to pass user-supplied input, which may lead to security vulnerabilities.

Reviewer 2: ```suggestion
      This function can execute arbitrary code. Calling it with
      user-supplied input may lead to security vulnerabilities.
```
More concise

      >>> eval('x+1')
      2

>    .. warning::

Reviewer 5: The `exec` warning is nice and prominent right under the function signature:

https://cpython-previews--118437.org.readthedocs.build/en/118437/library/functions.html#exec

This `eval` warning is halfway down, after an example:

https://cpython-previews--118437.org.readthedocs.build/en/118437/library/functions.html#eval

Shall we move this one up, under the signature, or perhaps right after the parameters table?

Author: Good question. Unfortunately `exec` has no parameters table so it is a bit inconsistent.

What do the others think, where should the warning be?

Reviewer 6: Personally I would like to see this warning under the parameters table for the ``eval`` and under the signature for the ``exec`` (since ``exec`` does not has parameter table).

Author: I've moved it now for `eval` as suggested (at least I think I did it correctly).

.. function:: exec(source, /, globals=None, locals=None, *, closure=None)

   .. warning::

>       Because this function can be used for executing arbitrary code, it is not recommended
>       to pass user-supplied input, which may lead to security vulnerabilities.

Reviewer 2: ```suggestion
      This function can execute arbitrary code. Calling it with
      user-supplied input may lead to security vulnerabilities.
```
